Provide Auth system for using key secret pair

openwarrior opened the pull request:
At present this client use  Access Tokens which restrict use of secret store.
To be able to use secret store by providing minimum changes AuthzModule
is Implemented. It performs all related work of generating new key and secret pair, Storing them as account, retrieving stored account, deleting stored account. 

ModifiedAuthzModule - Implement AuthzModule from Aerogear Authz
By this rest of connectivity remain same.

ModuleKeeper - Retain Modified AuthzModule in a Map in the imitation of
AuthorizationManager

Commit:
Provide Auth system for using key secret pair

ModifiedAuthzModule - Implement AuthzModule from Aerogear Authz
By this rest of connectivity remain same.

ModuleKeeper = Retain Modified AuthzModule in the imitation of
AuthorizationManager

pilhuhn commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+ * Retrieving stored account, Deleting stored account.
+ */
+
+public class ModifiedAuthzModule implements AuthzModule {
```
There should be a better class name like `HawkularAuthzModule` or `OAuth2AuthzModule` or such -- depending on what the addition to the base module is. Or perhaps `SecretStoreAuthzModule` -- @jpkrohling perhaps also has an idea.

jpkrohling replied:
+1 for `SecretStoreAuthzModule`

openwarrior replied:
SecretStoreAuthzModule got nominated  :smile:

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+
+public class ModifiedAuthzModule implements AuthzModule {
+
+    Context applicationContext;
```
Can (and should be) `private` and `final`.

openwarrior replied:
That right thanks

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+public class ModifiedAuthzModule implements AuthzModule {
+
+    Context applicationContext;
+    private SQLStore<OAuth2AuthzSession> sessionStore;
```
Can be `final`.

openwarrior replied:
done

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+    private SQLStore<OAuth2AuthzSession> sessionStore;
+
+    public ModifiedAuthzModule(Context context) {
+        applicationContext = context;
```
`applicationContext` can be renamed to `context`. Also, prefer this — `this.context = context.getApplicationContext()`.

openwarrior replied:
turning name to context. But as i am already sending applicationContext to it. Can't find need of 
this.context = context.getApplicationContext().

arturdryomov replied:
You shouldn’t rely on any conventions when building API. Enforce all you can instead of relying on rules. In this case it is simple, safe and cheap.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+
+    @Override
+    public boolean hasCredentials() {
+        OAuth2AuthzSession storedAccount = sessionStore.read("hawkular");
```
Please don’t ever, ever, ever use hardcoded text. Move it to a constant.

openwarrior replied:
Note is taken :)

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+        if (storedAccount != null) {
+            return true;
+        }
+        return false;
```
Can be simplified to `storedAccount != null`.

openwarrior replied:
Was having a sense that doing something Dumb

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+    public void addAccount(OAuth2AuthzSession session) {
+        sessionStore.save(session);
+    }
+
```
Redundant empty lines.

openwarrior replied:
Always get confused with this. Can you please point me to a convention about it.

arturdryomov replied:
Just have a single line between methods. As far as I know Hawkular doesn’t have any strict rules except Checstyle ones. You can take a look at [Google ones](https://google.github.io/styleguide/javaguide.html) for some motivation and inspiration.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+        return false;
+    }
+
+    private SQLStore<OAuth2AuthzSession> openSessionStore() {
```
Try to place methods in usage order, i. e. this one should be after the constructor.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+            }
+        } else {
+
+            AsyncTask<Void, Void, HeaderAndBody> task = new AsyncTask<Void, Void, HeaderAndBody>() {
```
Most likely you can get away with a simple thread or single threaded executor. This way you can avoid storing exceptions in fields.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+                    HeaderAndBody result;
+                    try {
+                        HttpProvider provider =
+                                new HttpRestProvider(new URL(((LoginActivity) (activity)).backendUrl.toString()
```
This casting is bad. Your class should be as abstract as possible allowing any activity to use it.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/ModifiedAuthzModule.java:
```diff
@@ -0,0 +1,197 @@
+                                + buildLoginData(((LoginActivity) (activity)).username,
+                                ((LoginActivity) (activity)).password));
+                        result = provider.post("");
+                        Log.d("Called with POST to: ", new String(provider.getUrl().toString()));
```
Use `Timber`. It will disable logging for release builds for you.

Commit:
Hardcoded Strings shifted to AuthEndPoints

-> overloaded requestAccess to avoid casting and more independent code

->renamed file and better conventions

openwarrior commented:
@ming13 @pilhuhn @jpkrohling  @danielpassos  Can you please review again. Applied approx everything except Thread coz wanted to leave more like cookbook code of aerogear. And I don't know if line redundancy is upto the standard or not but i tried.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+
+    @Override
+    public boolean hasCredentials() {
+        OAuth2AuthzSession storedAccount = sessionStore.read(AuthEndPoints.NAME);
```
Let’s simplify once again ;-) `return sessionStore.read(AuthEndpoints.NAME) != null`

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+
+public class SecretStoreAuthzModule implements AuthzModule {
+
+    private Context context;
```
Still not `final` for both fields. The empty line before them is redundant as well (sorry).

jpkrohling commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+                    provider.setDefaultHeader("Authorization", "Basic "
+                            + buildLoginData(username, password));
+                    result = provider.post("");
+                    Timber.e("Called with POST to: ", new String(provider.getUrl().toString()));
```
`Timber.e` -> sounds like "e" stands for "error", no? Shouldn't it be `Timber.d`?

arturdryomov replied:
Yep, `Timber.e` indicates error level posted to logcat.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+
+    public SecretStoreAuthzModule(Context context) {
+        this.context = context.getApplicationContext();
+        sessionStore = openSessionStore();
```
Personally I prefer to set all fields using `this.` in constructors, but you call the shots.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+    public void requestAccess(final Activity activity, final Callback<String> callback,
+                              final String username, final String password, final URL url) {
+
+        Exception exception;
```
I’m pretty sure that these two variables are useless, because they are duplicated in the inner class.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+    @Override
+    public void requestAccess(final Activity activity, final Callback<String> callback) {
+        OAuth2AuthzSession storedAccount = sessionStore.read(AuthEndPoints.NAME);
+        final String token = storedAccount.getAccessToken();
```
`final` is useless here. Also, consider this.

``` java
try {
    callback.onSuccess(sessionStore.read(AuthEndpoints.NAME).getAccessToken());
} catch (Exception e) {
    callback.onFailure(e);
}
```

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+        final String token = storedAccount.getAccessToken();
+        try {
+            callback.onSuccess(token);
+        } catch (Exception e) {
```
Generally it is a bad practice to catch every exception there is. Is there any reasoning behind this?

openwarrior replied:
Can't know list of multiple exception which may arise. May be caused less of experience.
How to know. If net disconnect, wrong URL (that doesn't have hawkular), wrong password. And many more.

jpkrohling replied:
It's a good practice to catch only exceptions you care about, ie, those that you are able to handle. Catching the generic "Exception" means you are handling everything. In general, only code that interfaces with the end user needs to have such generic catches.

On this case, you could handle some scenarios: if the connection fails mid-flight, you might want to retry it. If you got a 404, you might want to give up immediately, as the outcome won't change anyway. Similarly, if you got a "Forbidden", it might mean that the user's password is wrong. For all cases you didn't predict, let it explode until some upper layer catches it and handle it appropriately (for instance, asking the user to check the password, or try again later, or ...).

openwarrior replied:
@jpkrohling Problem arising here is we can't or actually don't want to do much about catch exception. It is just to be sent back via callback and handle there By showing toast or snackbar to the user which corresponding action he should take.

Even If I do something here with the catch It finally have to send whole of exception back to callback and again to be handled by callback onFailure(Exception e)

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+                    result = provider.post("");
+                    Timber.e("Called with POST to: ", new String(provider.getUrl().toString()));
+                    Timber.e("Result of the POST: ", new String(result.getBody()));
+                    json = new String(result.getBody());
```
As far as I see it you don’t need `json` variable at all.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/AuthEndPoints.java:
```diff
@@ -0,0 +1,30 @@
+ */
+package org.hawkular.client.android.auth;
+
+public final class AuthEndPoints {
```
Pretty sure it is a single word `Endpoints`.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/AuthEndPoints.java:
```diff
@@ -0,0 +1,30 @@
+    public static final String NAME = "hawkular";
+    public static final String STORE = "sessionStore";
+
+    public static final class Endpoints {
```
At this point it is mostly paths, not endpoints.

arturdryomov commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,185 @@
+        return storedAccount != null;
+    }
+
+    public void requestAccess(final Activity activity, final Callback<String> callback,
```
Suddenly it seems like you don’t need `activity` after all.

openwarrior commented:
@ming13 please don't make me dance. :smile: 
I really got fond of you.

openwarrior commented:
@danielpassos Please review the code

openwarrior commented:
@ming13 @pilhuhn @jpkrohling  If code look satisfactory please accept the pull request so that I can move ahead with more contribution. 

Generic Exceptions are removed.

pilhuhn commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,186 @@
+            protected HeaderAndBody doInBackground(Void... params) {
+                HeaderAndBody result;
+                try {
+                    HttpProvider provider =
```
Just curious - why do we here use a HttpProvider and not one of the pipes? (bear with me if I am confusing things)

openwarrior replied:
Didn't wanted to mess with the code took from cookbook of aerogear.
But can if this one look bad

openwarrior replied:
@pilhuhn  Can retain HttpProvider?

pilhuhn replied:
@danielpassos Wdyt?

pilhuhn commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,186 @@
+                    provider.setDefaultHeader("Authorization", "Basic "
+                            + buildLoginData(username, password));
+                    result = provider.post("");
+                    Timber.d("Called with POST to: ", new String(provider.getUrl().toString()));
```
One Timber.d call is enough to show url and result (?)

openwarrior replied:
Thats right. Should i remove it with rebase or with next pull reqest?

pilhuhn replied:
Doing it in this pr is probably best, as this is self-contained then.

openwarrior replied:
@pilhuhn done

pilhuhn commented:
A question and a comment - otherwise LGTM :+1:

danielpassos commented:
Hey guys, I'd like to review that before merge. Please, give me some time. I'll try do it tomorrow morning

openwarrior commented:
@pilhuhn  Changed from AsyncTask To background Thread in thought of improvement. Please take a look to check its not got messed up.

danielpassos commented:
Hey @anuj1708 

First of all, sorry about my delay in review this. I have some question about this PR:
1. Why we need that? The app works without it?
2. What is the different between it and [HttpBasicAuthenticationModule](https://github.com/aerogear/aerogear-android-auth/blob/master/aerogear-android-auth/src/main/java/org/jboss/aerogear/android/authentication/basic/HttpBasicAuthenticationModule.java)?
3. You have created a new [AuthzModule](https://github.com/aerogear/aerogear-android-authz/blob/8e7520bec583ce1ad34ab9098334b795a7d7850f/aerogear-android-authz/src/main/java/org/jboss/aerogear/android/authorization/AuthzModule.java), but I can not see where it is/will be used

openwarrior commented:
@danielpassos  
No problem :)

1.1  We need at present to shift from connecting to Key Cloak for Access Token to using Secret Store (provide key secret pair in response of a POST request).
-- Even if we retain using access token still for new feature (QR for key/secret). Which have been added to Hawkular and to be added to android client.
1.2 At present app works without it. but not including this is limiting the functions

2 . Diff from  HttpBasicAuthenticationModule Cause its more about retaining the key secret pair for long time. and that too we never always needing to go through this to auth user as user may already have key secret  pair. which are just to be checked if right or not by any other means.
1. It will be fed to pipe for various REST requests to be performed by pipe. It needed one of the pipe module containing credentials. Shifting to (or adding key secret pair) resulted in its need.

plus. When I asked summers about need. he said its not exactly we can't so with existing code and said to extend AuthenticationModule. But as authz was already being used and have prettier code. So I considered using it .

Sorry for so long text. Was not getting ho to compress the points. :)

danielpassos commented:
Gotcha! It make me confused because basic auth is is  part of _auth_ but has you sad the project is using _authz_. Anyway, could you please add some code _using_ it, to we can see it really working?

danielpassos commented:
@secondsun Could you please take a look of that?

openwarrior commented:
@danielpassos  I have approx entire code ready but that is not good in looks at present. So don't want to send pul req with that.
Can i place that to my repository to show?

Commit:
finalising for preview

secondsun commented on src/main/java/org/hawkular/client/android/auth/SecretStoreAuthzModule.java:
```diff
@@ -0,0 +1,158 @@
+        return sessionStore.read(AuthData.NAME) != null;
+    }
+
+    public void requestAccess(final Callback<String> callback,
```
You shouldn't be doing this.

You should be either a) using an authModule which has support for passing in a UN and PW or b) delegating the colleciton of the UN and PW to an intent and collecting the necessary tokens in a callback similar to how the intent based Oauth2AuthzModule works.

openwarrior commented:
I am needing to close this request. will work on it locally and resubmit. Thanks
